nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py

```diff
@@ -0,0 +1,416 @@
+        decoder_tokenizer_model = self.register_artifact(
+            "decoder_tokenizer.tokenizer_model", self._cfg.decoder_tokenizer.get('tokenizer_model')
+        )
+
+        self.encoder_tokenizer, self.decoder_tokenizer = MTEncDecModel.setup_enc_dec_tokenizers(
+            encoder_tokenizer_library=self.encoder_tokenizer_library,
+            encoder_tokenizer_model=encoder_tokenizer_model,
+            encoder_bpe_dropout=self._cfg.encoder_tokenizer.get('bpe_dropout', 0.0)
+            if self._cfg.encoder_tokenizer.get('bpe_dropout', 0.0) is not None
+            else 0.0,
+            encoder_model_name=None,
+            encoder_r2l=self._cfg.encoder_tokenizer.get('r2l', False),
+            decoder_tokenizer_library=self.decoder_tokenizer_library,
+            encoder_tokenizer_vocab_file=self._cfg.encoder_tokenizer.get('vocab_file', None),
+            decoder_tokenizer_model=decoder_tokenizer_model,
+            decoder_bpe_dropout=self._cfg.decoder_tokenizer.get('bpe_dropout', 0.0)
+            if self._cfg.decoder_tokenizer.get('bpe_dropout', 0.0) is not None
+            else 0.0,
+            decoder_model_name=None,
+            decoder_r2l=self._cfg.decoder_tokenizer.get('r2l', False),
+            special_tokens=self.special_tokens,
+        )
+
+        # Set up pre and post processors as well.
+        if self.multilingual:
+            (
+                self.source_processor_list,
+                self.target_processor_list,
+                self.multilingual_ids,
+            ) = MTEncDecModel.setup_multilingual_ids_and_processors(
+                src_language=self.src_language,
+                tgt_language=self.tgt_language,
+                tokenizer=self.encoder_tokenizer,  # Multilingual training requires shared tokenizers.
+                tokenizer_library=self.encoder_tokenizer_library,
+            )
+        else:
+            # After this call, the model will have  self.source_processor and self.target_processor objects
+            MTEncDecModel.setup_pre_and_post_processing_utils(
+                self.src_language, self.tgt_language, self.encoder_tokenizer_library, self.decoder_tokenizer_library,
+            )
+            self.multilingual_ids = [None]
+
+    def _build_vocab(self):
+        self.padded_vocab_size = self._vocab_size_with_padding(
+            orig_vocab_size=self.encoder_tokenizer.vocab_size,
+            make_vocab_size_divisible_by=self._cfg.get('make_vocab_size_divisible_by', 128),
+            tensor_model_parallel_size=self._cfg.get('tensor_model_parallel_size', 1),
+        )
+
+    def training_step(self, batch, batch_idx):
+        # Need to squeze dim 0 for tarred datasets since things are pre-batched and we ask the dataloader for batch size 1.
+        batch = [x.squeeze(dim=0) if x.ndim == 3 else x for x in batch]
+        hashitem = batch[0].sum().cpu().numpy().tolist()
+        assert hashitem not in self.train_hash_buffer
+        self.train_hash_buffer.add(hashitem)
```

Comment:
Why do we need to hash the training sets in order to verify uniqueness? Is this required or was left as part of debugging?

Reply:
Sorry, this needs to be removed. It was part of debugging. Done.